File: backend/models/Pregnancy.js
const prisma = require('../lib/prisma');

class Pregnancy {
  constructor(data) {
    this.id = data.id;
    this.dueDate = new Date(data.dueDate);
    this.lastMenstrualPeriod = new Date(data.lastMenstrualPeriod);
    this.notes = data.notes;
    this.createdAt = data.createdAt ? new Date(data.createdAt) : new Date();
    this.updatedAt = data.updatedAt ? new Date(data.updatedAt) : new Date();
  }

  // Calculate current week of pregnancy
  getCurrentWeek() {
    const now = new Date();
    const lmp = new Date(this.lastMenstrualPeriod);
    const daysSinceLMP = Math.floor((now - lmp) / (1000 * 60 * 60 * 24));
    return Math.floor(daysSinceLMP / 7) + 1;
  }

  // Calculate days until due date
  getDaysUntilDueDate() {
    const now = new Date();
    const dueDate = new Date(this.dueDate);
    return Math.ceil((dueDate - now) / (1000 * 60 * 60 * 24));
  }

  // Get current trimester
  getCurrentTrimester() {
    const week = this.getCurrentWeek();
    if (week <= 12) return 1;
    if (week <= 28) return 2;
    return 3;
  }

  // Get progress percentage
  getProgressPercentage() {
    const lmp = new Date(this.lastMenstrualPeriod);
    const dueDate = new Date(this.dueDate);
    const now = new Date();
    const totalDays = Math.floor((dueDate - lmp) / (1000 * 60 * 60 * 24));
    const elapsedDays = Math.floor((now - lmp) / (1000 * 60 * 60 * 24));
    return Math.min(Math.max(elapsedDays / totalDays, 0), 1);
  }

  async save() {
    const data = {
      dueDate: new Date(this.dueDate),
      lastMenstrualPeriod: new Date(this.lastMenstrualPeriod),
      notes: this.notes,
    };

    if (this.id) {
      // Update existing pregnancy
      const updated = await prisma.pregnancyData.update({
        where: { id: this.id },
        data: data,
      });
      return new Pregnancy(updated);
    } else {
      // Create new pregnancy
      const created = await prisma.pregnancyData.create({
        data: data,
      });
      return new Pregnancy(created);
    }
  }

  static async getCurrent() {
    const pregnancy = await prisma.pregnancyData.findFirst({
      orderBy: { createdAt: 'desc' },
    });
    
    if (pregnancy) {
      return new Pregnancy(pregnancy);
    }
    
    return null;
  }

  static async delete(id) {
    await prisma.pregnancyData.delete({
      where: { id },
    });
  }

  toJSON() {
    return {
      id: this.id,
      dueDate: this.dueDate,
      lastMenstrualPeriod: this.lastMenstrualPeriod,
      notes: this.notes,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
      currentWeek: this.getCurrentWeek(),
      daysUntilDueDate: this.getDaysUntilDueDate(),
      currentTrimester: this.getCurrentTrimester(),
      progressPercentage: this.getProgressPercentage()
    };
  }
}

module.exports = Pregnancy;

File: luma/lib/models/pregnancy.dart
import '../services/device_timezone_service.dart';

class Pregnancy {
  final String id;
  final DateTime dueDate;
  final DateTime lastMenstrualPeriod;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? notes;

  Pregnancy({
    required this.id,
    required this.dueDate,
    required this.lastMenstrualPeriod,
    required this.createdAt,
    required this.updatedAt,
    this.notes,
  });

  // Calculate current week of pregnancy
  int get currentWeek {
    final now = DeviceTimezoneService.now();
    final daysSinceLMP = now.difference(lastMenstrualPeriod).inDays;
    return (daysSinceLMP / 7).floor() + 1;
  }

  // Calculate days until due date
  int get daysUntilDueDate {
    final now = DeviceTimezoneService.now();
    return dueDate.difference(now).inDays;
  }

  // Get current trimester
  int get currentTrimester {
    if (currentWeek <= 12) return 1;
    if (currentWeek <= 28) return 2;
    return 3;
  }

  // Get progress percentage
  double get progressPercentage {
    final totalDays = dueDate.difference(lastMenstrualPeriod).inDays;
    final elapsedDays = DeviceTimezoneService.now().difference(lastMenstrualPeriod).inDays;
    return (elapsedDays / totalDays).clamp(0.0, 1.0);
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'dueDate': dueDate.toIso8601String(),
      'lastMenstrualPeriod': lastMenstrualPeriod.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'notes': notes,
    };
  }

  factory Pregnancy.fromJson(Map<String, dynamic> json) {
    return Pregnancy(
      id: json['id'] ?? '',
      dueDate: DateTime.parse(json['dueDate']),
      lastMenstrualPeriod: DateTime.parse(json['lastMenstrualPeriod']),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      notes: json['notes'],
    );
  }

  Pregnancy copyWith({
    String? id,
    DateTime? dueDate,
    DateTime? lastMenstrualPeriod,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? notes,
  }) {
    return Pregnancy(
      id: id ?? this.id,
      dueDate: dueDate ?? this.dueDate,
      lastMenstrualPeriod: lastMenstrualPeriod ?? this.lastMenstrualPeriod,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      notes: notes ?? this.notes,
    );
  }
}
